src/txrequest.h

```diff
@@ -0,0 +1,342 @@
+    {
+        uint64_t low_bits = 0;
+        if (!first) {
+            low_bits = CSipHasher(m_k0, m_k1).Write(txid.begin(), txid.size()).Write(peer).Finalize() >> 1;
+        }
+        return low_bits | uint64_t(inbound) << 63;
+    }
+};
+
+/** Data structure to keep track of, and schedule, transaction downloads from peers.
+ *
+ * === High level behavior ===
+ *
+ * Transaction downloads are attempted in random order, first choosing among outbound peers, then inbound ones.
+ * Inbound peers are only considered after a (configurable) inbound delay, to give time to outbound peers. When a
+ * NOTFOUND is received, an invalid transaction is received, or a download times out (after a configurable delay),
+ * the next candidate is immediately scheduled according to the same rules above. The same transaction is never
+ * re-requested from the same peer, unless the transaction was forgotten about in the mean time. This happens
+ * whenever no candidates remain, or when a transaction is succesfully received.
+ *
+ *
+ * === Specification ===
+ *
+ * The data structure maintains a collection of entries:
+ *
+ * - CANDIDATE (txid, peer, reqtime) entries representing txids that were announced by peer, and become available for
+ *   download after reqtime has passed (which is immediately for outbound peers, and inbound_delay after announcement
+ *   for inbound ones).
+ *
+ * - REQUESTED (txid, peer, exptime) entries representing txids that have been requested, and we're awaiting a response
+ *   for from that peer. The exptime value determines when the request times out.
+ *
+ * - COMPLETED (txid, peer) entries representing txids that have been requested from a peer, but they timed out, a
+ *   NOTFOUND message was received for them, or an invalid response was received. They're only kept around to prevent
+ *   downloading them again. If only COMPLETED entries exist for a given txid remain (so no CANDIDATE or REQUESTED
+ *   ones), all of them are deleted (this is an invariant, and maintained by all operations below).
+ *
+ * The following operations are supported on this data structure:
+ *
+ * - ReceivedInv(txid, peer) adds a new CANDIDATE entry, unless one already exists for that (txid, peer) combination
+ *   (whether it's CANDIDATE, REQUESTED, or COMPLETED).
+ *
+ * - DeletedPeer(peer) deletes all entries for a given peer. It should be called when a peer goes offline.
+ *
+ * - AlreadyHaveTx(txid) deletes all entries for a given txid. It should be called when a transaction is succesfully
+ *   added to the mempool, seen in a block, or for whatever reason we no longer care about it.
+ *
+ * - ReceivedResponse(peer, txid) converts any CANDIDATE or REQUESTED entry to a COMPLETED one. It should be called
+ *   whenever a transaction or NOTFOUND was received from a peer. When the transaction is acceptable, AlreadyHaveTx
+ *   should be called instead of (or in addition to) this operation.
+
```

Comment:
missing `*`. I don't know if this messes with doxygen parsing.

Reply:
Fixed.